secure/json-rpc/src/lib.rs

```diff
@@ -0,0 +1,595 @@
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq)]
+pub struct Bytes(pub String);
+
+impl Bytes {
+    pub fn into_bytes(self) -> Result<Vec<u8>, Error> {
+        Ok(hex::decode(self.0)?)
+    }
+}
+
+impl From<&[u8]> for Bytes {
+    fn from(bytes: &[u8]) -> Self {
+        Self(hex::encode(bytes))
+    }
+}
+
+impl From<&Vec<u8>> for Bytes {
+    fn from(bytes: &Vec<u8>) -> Self {
+        Self(hex::encode(bytes))
+    }
+}
+
+/// Testing infrastructure for the lightweight JSON client above. Hooks into the original JSON RPC
+/// infrastructure.
+#[cfg(test)]
+mod test {
+    use crate::JsonRpcClient;
+    use anyhow::Result;
+    use futures::{channel::mpsc::channel, StreamExt};
+    use libra_config::utils;
+    use libra_crypto::{ed25519::Ed25519PrivateKey, HashValue, PrivateKey, Uniform};
+    use libra_json_rpc::bootstrap;
+    use libra_types::{
+        account_address::AccountAddress,
+        account_config::{AccountResource, BalanceResource},
+        account_state::AccountState,
+        account_state_blob::{AccountStateBlob, AccountStateWithProof},
+        contract_event::ContractEvent,
+        event::{EventHandle, EventKey},
+        ledger_info::LedgerInfoWithSignatures,
+        mempool_status::{MempoolStatus, MempoolStatusCode},
+        proof::{
+            AccountStateProof, AccumulatorConsistencyProof, AccumulatorProof, SparseMerkleProof,
+        },
+        test_helpers::transaction_test_helpers::get_test_signed_txn,
+        transaction::{
+            SignedTransaction, TransactionInfo, TransactionListWithProof, TransactionWithProof,
+            Version,
+        },
+        validator_change::ValidatorChangeProof,
+        vm_error::StatusCode,
+    };
+    use libradb::errors::LibraDbError::NotFound;
+    use std::{collections::BTreeMap, convert::TryFrom, sync::Arc};
+    use storage_interface::DbReader;
+    use storage_proto::StartupInfo;
+    use tokio::runtime::Runtime;
+    use vm_validator::{
+        mocks::mock_vm_validator::MockVMValidator, vm_validator::TransactionValidation,
+    };
+
+    #[test]
+    fn test_submit_transaction() {
+        let mock_db = create_empty_mock_db();
+        let (client, _server) = create_client_and_server(mock_db, true);
+        let signed_transaction = generate_signed_transaction();
+
+        // Ensure transaction submitted and validated successfully
+        let result = client.submit_signed_transaction(signed_transaction);
+        assert!(result.is_ok());
+    }
+
+    #[test]
+    fn test_submit_transaction_failure() {
+        let mock_db = create_empty_mock_db();
+        // When creating the JSON RPC server, specify 'mock_validator=false' to prevent a vm validator
+        // being passed to the server. This will cause any transaction submission requests to the JSON
+        // RPC server to fail, thus causing the server to return an error.
+        let (client, _server) = create_client_and_server(mock_db, false);
```

Comment:
if this is the only case you pass a vm_validator, why not just mutate the signed_transaction to be a bad signed transaction? that way you don't need this modified vm_validator?

Reply:
I did originally try this, but (iirc) even with the badly signed transaction, the call still fails without passing in a vm_validator.. Going to leave it as-is for now!